ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,358 @@
+    return getCurrentCycleLog().compactionDetails;
+  }
+
+  /**
+   * @return the {@link StoreFindToken} until which data has been copied and flushed.
+   */
+  StoreFindToken getSafeToken() {
+    return getCurrentCycleLog().safeToken;
+  }
+
+  /**
+   * Sets the {@link StoreFindToken} until which data is copied and flushed.
+   * @param safeToken the {@link StoreFindToken} until which data is copied and flushed.
+   */
+  void setSafeToken(StoreFindToken safeToken) {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getPhase().equals(Phase.COPY)) {
+      throw new IllegalStateException("Cannot set a safe token - not in COPY phase");
+    }
+    cycleLog.safeToken = safeToken;
+    flush();
+  }
+
+  /**
+   * Marks the start of the copy phase.
+   */
+  void markCopyStart() {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getPhase().equals(Phase.PREPARE)) {
+      throw new IllegalStateException("Should be in PREPARE phase to transition to COPY phase");
+    }
+    cycleLog.copyStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the switch phase.
+   */
+  void markSwitchStart() {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getPhase().equals(Phase.COPY)) {
+      throw new IllegalStateException("Should be in COPY phase to transition to SWITCH phase");
+    }
+    cycleLog.switchStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the cleanup phase.
+   */
+  void markCleanupStart() {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getPhase().equals(Phase.SWITCH)) {
+      throw new IllegalStateException("Should be in SWITCH phase to transition to CLEANUP phase");
+    }
+    cycleLog.cleanupStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the current compaction cycle as complete.
+   */
+  void markCycleComplete() {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getPhase().equals(Phase.CLEANUP)) {
+      throw new IllegalStateException("Should be in CLEANUP phase to complete cycle");
+    }
+    cycleLog.cycleEndTime = time.milliseconds();
+    currentIdx++;
+    flush();
+  }
+
+  /**
+   * Closes the compaction log. If compaction is complete, renames the log to keep a permanent record.
+   */
+  @Override
+  public void close() {
+    if (file.exists() && getCompactionPhase().equals(Phase.DONE)) {
+      String dateString = new Date(startTime).toString();
+      File savedLog =
+          new File(file.getAbsolutePath() + BlobStore.SEPARATOR + startTime + BlobStore.SEPARATOR + dateString);
+      if (!file.renameTo(savedLog)) {
+        throw new IllegalStateException("Compaction log could not be renamed after completion of compaction");
+      }
+    }
+  }
+
+  /**
+   * @return the {@link CycleLog} for the current compaction cycle.
+   */
+  private CycleLog getCurrentCycleLog() {
+    if (currentIdx >= cycleLogs.size()) {
```

Comment:
when the last cycle is complete, currentIdx is equal to cycleLogs.size() right. So, this method assumes that no one should call into this method, when last cycle is complete?